sympy/solvers/solveset.py

```diff
@@ -798,7 +798,11 @@ def _solve_as_poly(f, symbol, domain=S.Complexes):
     """
     result = None
     if f.is_polynomial(symbol):
-        solns = roots(f, symbol, cubics=True, quartics=True,
+        if (domain.is_subset(S.Reals) or domain==S.Reals) and (f.atoms(Symbol)==f.free_symbols) and (FiniteSet(*Poly(f,symbol).coeffs()).is_subset(S.Rationals)):
+            solns=FiniteSet(*real_roots(Poly(f,symbol))).evalf()
+            return solns
+        else:
+            solns = roots(f, symbol, cubics=True, quartics=True,
```

Comment:
I would write this like:
```python
if f.is_polynomial(symbol):
    f_poly = f.as_poly(Symbol)
    if f_poly.domain.is_QQ:
        if domain.is_subset(S.Reals):
            result = FiniteSet(*f_poly.real_roots())
        else:
            result = FiniteSet(*f_poly.all_roots())        
    else:
        solns = roots(f_poly, ...)
        ...
```
All branches will ultimately end up converting to Poly so we might as well do that at the start and then the poly domain tells us if the coefficients are rational more easily.

It would be better in general though if this was added as an option to the `roots` function so you can just do something like:
```python
solns = roots(f, symbol, rootof=True, ...)
```
This is not the only place that would benefit from being able to call `roots` and have it return `RootOf` when possible.

Reply:
I'll look into the failed tests and try to solve the issue. Thanks a lot for the advice on improving the code and all the help.

Reply:
I believe this PR may be closed now as #24904 supersedes this in functionality. Would love to contribute if more changes can be made (if any are left out by your PR).

Reply:
No gh-24904 does not supersede the changes here. The intention there is just to make the roots function more maintainable which would make it easier to add something like the `rootof` parameter that I mentioned above. For now though that will not solve the problems that this PR aims to solve.